web/src/narrow.js

```diff
@@ -405,93 +421,135 @@ export function activate(raw_terms, opts) {
+            // marked some messages as unread in the last All messages session and thus prevented reading.
+            message_lists.current.resume_reading();
+            // Reset the collapsed status of messages rows.
+            condense.condense_and_collapse(message_lists.current.view.$list.find(".message_row"));
+            message_edit.handle_narrow_deactivated();
+            widgetize.set_widgets_for_list();
+            message_feed_top_notices.update_top_of_narrow_notices(msg_list);
+
+            // We may need to scroll to the selected message after swapping
+            // the currently displayed center panel to All messages.
+            message_viewport.maybe_scroll_to_selected();
+        } else {
+            if (id_info.target_id === id_info.final_select_id) {
+                then_select_offset = opts.then_select_offset;
             }
 
-            message_fetch.load_messages_for_narrow({
-                anchor,
-                cont() {
-                    if (!select_immediately) {
-                        render_message_list_with_selected_message({
-                            id_info,
-                            select_offset: then_select_offset,
-                            msg_list: message_lists.current,
-                        });
-                    }
-                },
-                msg_list,
-            });
+            select_immediately = id_info.local_select_id !== undefined;
+
+            {
+                let anchor;
+
+                // Either we're trying to center the narrow around a
+                // particular message ID (which could be max_int), or we're
+                // asking the server to figure out for us what the first
+                // unread message is, and center the narrow around that.
+                switch (id_info.final_select_id) {
+                    case undefined:
+                        anchor = "first_unread";
+                        break;
+                    case -1:
+                        // This case should never happen in this code path; it's
+                        // here in case we choose to extract this as an
+                        // independent reusable function.
+                        anchor = "oldest";
+                        break;
+                    case LARGER_THAN_MAX_MESSAGE_ID:
+                        anchor = "newest";
+                        break;
+                    default:
+                        anchor = id_info.final_select_id;
+                }
+                message_fetch.load_messages_for_narrow({
+                    anchor,
+                    cont() {
+                        if (!select_immediately) {
+                            render_message_list_with_selected_message({
+                                id_info,
+                                select_offset: then_select_offset,
+                                msg_list: message_lists.current,
+                                ...select_opts,
```

Comment:
I think it would probably feel nicer if `select_opts` were an independent parameter.